sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala

```diff
@@ -138,39 +137,38 @@ case class SessionHolder(userId: String, sessionId: String, session: SparkSessio
   }
 
   /**
-   * Add ExecuteHolder to this session.
+   * Add an operation ID to this session.
    *
-   * Called only by SparkConnectExecutionManager under executionsLock.
+   * Called only by SparkConnectExecutionManager when a new execution is started.
    */
-  @GuardedBy("SparkConnectService.executionManager.executionsLock")
-  private[service] def addExecuteHolder(executeHolder: ExecuteHolder): Unit = {
+  private[service] def addOperationId(operationId: String): Unit = {
     if (closedTimeMs.isDefined) {
       // Do not accept new executions if the session is closing.
       throw new SparkSQLException(
         errorClass = "INVALID_HANDLE.SESSION_CLOSED",
         messageParameters = Map("handle" -> sessionId))
     }
 
-    val oldExecute = executions.putIfAbsent(executeHolder.operationId, executeHolder)
-    if (oldExecute != null) {
-      // the existence of this should alrady be checked by SparkConnectExecutionManager
-      throw new IllegalStateException(
-        s"ExecuteHolder with opId=${executeHolder.operationId} already exists!")
+    var existed = true
+    operationIds.computeIfAbsent(
```

Comment:
How about this:
```scala
val alreadyExists = operationIds.putIfAbsent(operationId, ()) != null
```
That IMO is simpler and more readable.

Reply:
(basically the old code)

Reply:
Sadly, the compiler prohibits comparison between `Unit` and `null`; + Option[Unit](null) gives `Some(())`.

Reply:
would `val alreadyExists = operationIds.putIfAbsent(operationId, ()) != Some(())`  work?

Reply:
Good idea, but unfortunately, doesn't work: warning.

[error] /Users/changgyoo.park/spark/sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala:153:67: comparing values of types Unit and Some[Unit] using `==` will always yield false